azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -1002,12 +1003,30 @@ private void pollExecution() {
               .selectAndUpdateExecution(this.executorId, FlowRunnerManager.this.active);
           if (execId != -1) {
             FlowRunnerManager.logger.info("Submitting flow " + execId);
-            submitFlow(execId);
-            FlowRunnerManager.this.commonMetrics.markDispatchSuccess();
+            try {
+              submitFlow(execId);
+              FlowRunnerManager.this.commonMetrics.markDispatchSuccess();
+              this.numRetries = 0;
+            } catch (final ExecutorManagerException e) {
+              // If the flow fails to be submitted, then unset its executor id in DB so that other
+              // executors can pick up this flow and submit again.
+              FlowRunnerManager.this.executorLoader.unsetExecutorIdForExecution(execId);
+              throw new ExecutorManagerException(
+                  "Unset executor id " + this.executorId + " for execution " + execId, e);
+            }
           }
         } catch (final Exception e) {
           FlowRunnerManager.logger.error("Failed to submit flow ", e);
           FlowRunnerManager.this.commonMetrics.markDispatchFail();
+          this.numRetries = this.numRetries + 1;
+          try {
+            // Implement exponential backoff retries when flow submission fails,
+            // i.e., sleep 1s, 2s, 4s, 8s ... before next retries.
+            Thread.sleep((long) (Math.pow(2, this.numRetries) * 1000));
```

Comment:
As suggested by @ypadron-in , the backoff time should be implemented inside the scheduler itself. But I didn't find out any existing library that provides this functionality, so just implemented my own backoff time during each scheduled run. Will keep it in the todo list.